test/e2e/auth/subjectreviews.go

```diff
@@ -0,0 +1,173 @@
+/*
+Copyright 2022 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package auth
+
+import (
+	"context"
+	"fmt"
+
+	authorizationv1 "k8s.io/api/authorization/v1"
+	v1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	utilrand "k8s.io/apimachinery/pkg/util/rand"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/rest"
+	"k8s.io/kubernetes/test/e2e/framework"
+	e2epod "k8s.io/kubernetes/test/e2e/framework/pod"
+	admissionapi "k8s.io/pod-security-admission/api"
+	"k8s.io/utils/pointer"
+
+	"github.com/onsi/ginkgo/v2"
+	imageutils "k8s.io/kubernetes/test/utils/image"
+)
+
+var _ = SIGDescribe("SubjectReview", func() {
+	f := framework.NewDefaultFramework("subjectreview")
+	f.NamespacePodSecurityEnforceLevel = admissionapi.LevelPrivileged
+
+	ginkgo.It("should support SubjectReview API operations", func() {
+
+		AuthClient := f.ClientSet.AuthorizationV1()
+		ns := f.Namespace.Name
+
+		podClient := f.ClientSet.CoreV1().Pods(ns)
```

Comment:
A few suggestions:

* don't create a pod... make a more light-weight API call like listing configmaps
* build the full username/groups/uid for the serviceaccount using the utility methods serviceaccount.MakeUsername and serviceaccount.MakeGroupNames (and include "system:authenticated" in the serviceaccount groups)
* use the full username/groups/uid when impersonating and performing the subjectaccessreview

For example:

```
// make a serviceaccount
sa, err := f.ClientSet.CoreV1().ServiceAccounts(ns).Create(...)

// build the username and groups
saUsername := serviceaccount.MakeUsername(sa.Namespace, sa.Name)
saGroups := []string{"system:authenticated"}
saGroups = append(saGroups, serviceaccount.MakeGroupNames(sa.Namespace)...)
...

// run a subject access review
sar := &authorizationv1.SubjectAccessReview{
	Spec: authorizationv1.SubjectAccessReviewSpec{
		ResourceAttributes: &authorizationv1.ResourceAttributes{
			Verb:      "list",
			Resource:  "configmaps",
			Namespace: ns,
			Version:   "v1",
		},
		User: saUsername,
		Groups: saGroups,
		UID: sa.UID,
	},
}
...

// make impersonated config that matches the service account
config := rest.CopyConfig(f.ClientConfig())
config.Impersonate = rest.ImpersonationConfig{
	UserName: saUsername,
	UID: sa.UID,
	Groups: saGroups,
}
...
expectedAllowed = sarResponse.Status.Allowed

// make a list configmaps request
_, err = impersonatedClientSet.CoreV1().ConfigMaps(ns).List(context.TODO(), metav1.ListOptions{})
actuallyAllowed := false
switch {
case apierrors.IsForbidden(err):
  actuallyAllowed = false
case err != nil:
  ... other unexpected error, fail the test
default:
  actuallyAllowed = true
}

if actuallyAllowed != expectedAllowed {
  ... fail the test
}
```

Reply:
Thanks for the feedback and tips @liggitt  Ready for another review.